libs/application-generic/src/usecases/workflow/delete-workflow/delete-workflow.usecase.ts

```diff
@@ -0,0 +1,136 @@
+import { Injectable } from '@nestjs/common';
+
+import {
+  ControlValuesRepository,
+  MessageTemplateRepository,
+  NotificationTemplateEntity,
+  NotificationTemplateRepository,
+  PreferencesRepository,
+} from '@novu/dal';
+import { PreferencesTypeEnum, WorkflowOriginEnum } from '@novu/shared';
+
+import { DeleteWorkflowCommand } from './delete-workflow.command';
+import { InvalidateCacheService } from '../../../services/cache/invalidate-cache.service';
+import { GetWorkflowByIdsUseCase } from '../get-workflow-by-ids/get-workflow-by-ids.usecase';
+import { GetWorkflowByIdsCommand } from '../get-workflow-by-ids/get-workflow-by-ids.command';
+import {
+  buildNotificationTemplateIdentifierKey,
+  buildNotificationTemplateKey,
+} from '../../../services/cache/key-builders';
+
+@Injectable()
+export class DeleteWorkflowUseCase {
+  constructor(
+    private notificationTemplateRepository: NotificationTemplateRepository,
+    private messageTemplateRepository: MessageTemplateRepository,
+    private getWorkflowByIdsUseCase: GetWorkflowByIdsUseCase,
+    private preferencesRepository: PreferencesRepository,
+    private invalidateCache: InvalidateCacheService,
+    private controlValuesRepository: ControlValuesRepository,
+  ) {}
+
+  async execute(command: DeleteWorkflowCommand): Promise<void> {
+    const workflowEntity = await this.getWorkflowByIdsUseCase.execute(
+      GetWorkflowByIdsCommand.create({
+        ...command,
+        identifierOrInternalId: command.identifierOrInternalId,
+      }),
+    );
+
+    await this.deleteRelatedEntities(command, workflowEntity);
+  }
+
+  private async deleteRelatedEntities(
+    command: DeleteWorkflowCommand,
+    workflow: NotificationTemplateEntity,
+  ) {
+    await this.deleteControlValues(command, workflow);
+    await this.deleteMessageTemplates(workflow, command);
+    await this.deletePreferences(command, workflow);
+    await this.invalidateCacheForWorkflow(workflow, command);
```

Comment:
Please invalidate the cache first as its an isolated task that should be done regardless of what happens in Mongo. Note that cache invalidation can't be part of the transaction as it targets a different DB.

Reply:
done